content/blog/reducing-openapi.md

```diff
@@ -0,0 +1,252 @@
+        id:
+          type: integer
+        failed:
+          type: boolean
+        status:
+          type: string
+      additionalProperties: false
+    GetJobResponse:
+      type: object
+      properties:
+        id:
+          type: integer
+        failed:
+          type: boolean
+        status:
+          type: string
+      additionalProperties: false
+    UnauthorizedError:
+      type: object
+      properties:
+        detail:
+          type: string
+  securitySchemes:
+    basicAuth:
+      type: http
+      scheme: basic
+</code>
+</pre>
+</div>
+
+The JSON object hierarchy can be described in the following diagram:  
+(Some objects have been added for completeness)
+```
+openapi
+info
+├── title
+└── version
+servers
+security
+paths
+├── path_1
+└── path_2
+components
+├── schemas
+│   ├── shema_1
+│   └── schema_2
+├── securitySchemas
+├── parameters
+└── headers
+```
+ 
+The document begins by specifying the openAPI version, an info object, and specifying the servers.
+```
+openapi: 3.0.1
+info:
+ title: Kubernetes
+ version: v1.29.6+aba1e8d
+servers:
+ - url: /
+security:
+ - BearerToken: []
+ ```
+
+The document will then display the system's paths, and must include at least one path.  
+Under each path, the document specifies the path URI, an HTTP method, parameters (if needed), the requestBody, and defines the responses. 
+Note that both the request and the responses can reference a component from a different part of the document. This will be important when we try to reduce the document to our needed purposes. 
+
+<div style="height: 200px; overflow-x: auto;">
+<pre>
+<code>
+paths:
+/api/v2/job_templates/{job_template_id}/launch/:
+post:
+    description: Launch a job from a template id
+    operationId: launchJob
+    parameters:
+    - name: job_template_id
+        in: path
+        required: true
+        schema:
+        type: integer
+    requestBody:
+    content:
+        application/json:
+            schema:
+            $ref: '#/components/schemas/LaunchJobRequest'
+    responses:
+    "201":
+        description: Created
+        content:
+        application/json:
+            schema:
+            $ref: '#/components/schemas/LaunchJobResponse'
+</code>
+</pre>
+</div>
+
+
+Lastly, The document specifies the components it uses in its paths. 
+A component can be a schema for any resource, some parameter specification object, HTTP headers and more. To view more openAPI documents to get a sense of some practical examples, [see:](https://github.com/parodos-dev/serverless-workflow-examples/tree/main/aap-db-deploy-quarkus/src/main/resources/specs)
+```
+components:
+ schemas:
+   LaunchJobRequest:
+     type: object
+     properties:
+       limit:
+         type: string
+```
+
+
+Note the securitySchemas object, which specifies the authentication method used: 
+```
+securitySchemes:
+   BearerToken:
+     type: http
+     scheme: bearer
+```
+
+## How to reduce an OpenAPI Document to workable size?
```

Comment:
pls explore if there is an existing tool to do it. otherwise, we'll ask @gabriel-farache for the tool he created or create one.
after all, that logic will have to be implemented in the `kn-workflow` cli (or at some part of the workflow's development process).

Reply:
So about this, I played with some options (using yq / jq in scripts, using golang schemas, using python+json manipulation), nothing stood out to me as 'the best' for the tool, so I would like to see Gabriel approach and take it from there
@masayag Do you have any intuition on what would work the best?

Reply:
I committed the script in the aap-db branch that was created for some demo @masayag did: https://github.com/parodos-dev/serverless-workflows/blob/feat/aap-db-deploy/extract_from_swagger_spec.py 

```
python extract.py openapi_spec.json filtered_openapi_spec.json "/apis/apps/v1/namespaces/{namespace}/deployments post" "/apis/apps/v1/namespaces/{namespace}/deployments/{name} get" "/api/v1/namespaces/{namespace}/services post" "/apis/route.openshift.io/v1/namespaces/{namespace}/routes post" "/apis/route.openshift.io/v1/namespaces/{namespace}/routes/{name} get" "/apis/route.openshift.io/v1/namespaces/{namespace}/routes/{name}/status get"
``` 
I takes tuple (path, method) to extract from the spec and it takes all dependencies (refs) that come with it in a recursive way,  even so some are missed sometime.

Why swagger? Because that is what is returned by
```
oc get --raw /openapi/v2 > openapi-v2.json
```

Reply:
I provided a reference to that script to @ElaiShalevRH offline, however, pls note this script isn't complete and failed to work for some of the openapi spec files (iirc, when there was paratemeters/metadata and/or recursive reference).